_episodes/08-reproducibility.md

```diff
@@ -0,0 +1,144 @@
+---
+title: "Containers in research workflows: reproducibility and granularity"
+teaching: 20
+exercises: 0
+questions:
+- "How can I use container images to make my research more reproducible?"
+- "How do I incorporate containers into my research workflow?"
+- "What are container orchestration tools and how can they potentially help me?"
+objectives:
+- "Understand how container images can help make research more reproducible."
+- "Understand what practical steps I can take to improve the reproducibility of my research using containers."
+- "Know that container orchestration tools are and what they can do"
+keypoints:
+- "Container images allow us to encapsulate the computation (and data) we have used in our research."
+- "Using a service such as Docker Hub allows us to easily share computational work we have done."
+- "Using container images along with a DOI service such as Zenodo allows us to capture our work and enables reproducibility."
+- "Tools such as Docker Compose, Docker Swarm and Kubernetes allow us to describe how multiple containers work together."
+---
+
+Although this workshop is titled "Reproducible computational environments using containers", so far we have mostly covered the mechanics of using Docker with only passing reference the the reproducibility aspects. In this section, we discuss these aspects in more detail.
+
+> ## Work in progress...
+> Note that reproducibility aspects of software and containers are an active area of research, discussion and development so are subject to many changes. We will present some ideas and approaches here but best practices will likely evolve in the near future.
+{: .callout}
+
+## Reproducibility
+
+By *reproducibility* here we mean the ability of someone else (or your future self) being able to reproduce what you did computationally at a particular time (be this in research, analysis or something else) as closely as possible even if you do not have access to exactly the same hardware resources that you had when you did the original work.
+
+Hopefully it is reasonably self-evident why containers are an attractive technology to help with reproducibility:
+
+  - The same computational work can be run across multiple different technologies seamlessly (e.g. Windows, macOS, Linux).
+  - You can save the exact process that you used for your computational work (rather than relying on potentially incomplete notes).
+  - You can save the exact versions of software and their dependencies in the image.
+  - You can access legacy versions of software and underlying dependencies which may not be generally available any more.
```

Comment:
Maybe we should add another reason below this line about also keeping a copy of the data with the scripts and analysis in an image?